src/lambda_vm/compiler.ml

```diff
@@ -0,0 +1,134 @@
+open Ast
+open Ir
+open Checks
+
+module String_map = Map.Make (String)
+
+type error = (* user program bugs *)
+  | Undefined_variable
+exception Error of error
+let raise error = raise (Error error)
+
+let compile_prim prim =
+  match prim with
+  | Neg -> P_neg
+  | Add -> P_add
+  | Sub -> P_sub
+  | Mul -> P_mul
+  | Div -> P_div
+  | Rem -> P_rem
+  | And -> P_and
+  | Or -> P_or
+  | Xor -> P_xor
+  | Lsl -> P_lsl
+  | Lsr -> P_lsr
+  | Asr -> P_asr
+
+module Vars = Map_with_cardinality.Make (String)
+
+let burn_gas gas vars code =
+  (match code with
+  | Var _
+  | Lam _ ->
+    let cardinality = Vars.cardinal vars in
+    Gas.burn_log2 gas ~cardinality
+  | App _
+  | Const _
+  | Prim _
+  | If _
+  | Pair _
+  | Fst _
+  | Snd _ ->
+    Gas.burn_constant gas);
+  check_gas gas
+
+let rec compile_expr ~stack gas next_ident vars code =
+  let stack = stack - 1 in
+
+  check_stack ~stack;
+  burn_gas gas vars code;
+
+  let compile_expr vars code = compile_expr ~stack gas next_ident vars code in
+
+  match code with
+  | Var var -> (
+    match Vars.find var vars with
+    | Some ident -> E_var ident
+    | None -> raise Undefined_variable)
+  | Lam (var, body) ->
+    let ident =
+      let ident = !next_ident in
+      next_ident := Ident.next ident;
+      ident in
+    let vars = Vars.add var ident vars in
+    let body = compile_expr vars body in
+    E_lam (ident, body)
+  | App { funct; arg } ->
+    let funct = compile_expr vars funct in
+    let arg = compile_expr vars arg in
+    E_app { funct; arg }
+  (* prims *)
+  | Const value -> E_const value
+  | Prim prim ->
+    let prim = compile_prim prim in
+    E_prim prim
+  (* branching *)
+  | If { condition; then_; else_ } ->
+    let condition = compile_expr vars condition in
+    let then_ = compile_expr vars then_ in
+    let else_ = compile_expr vars else_ in
+    E_if { condition; then_; else_ }
+  (* memory *)
+  | Pair { left; right } ->
```

Comment:
In Haskell, `left` and `right` are related to the `Either` monad. I think `Either` also exist on OCaml (sadly not a Monad), aren't `left` and `right` confusing?
I feel like `Pair { left; right }` should be something like `Pair { first; second }` what do you think?

Reply:
Yes, those are better names, and this is correct. Especially because there is `E_fst` and `E_snd`